src/librepgp/stream-common.c

```diff
@@ -0,0 +1,445 @@
+        cache->pos = 0;
+    }
+
+    while (cache->len < len) {
+        read = src->read(src, &cache->buf[cache->len], sizeof(cache->buf) - cache->len);
+        if (read == 0) {
+            memcpy(buf, &cache->buf[0], cache->len);
+            return cache->len;
+        } else if (read < 0) {
+            return -1;
+        } else {
+            cache->len += read;
+            if (cache->len >= len) {
+                memcpy(buf, &cache->buf[0], len);
+                return len;
+            }
+        }
+    }
+
+    return -1;
+}
+
+ssize_t
+src_skip(pgp_source_t *src, size_t len)
+{
+    ssize_t res;
+    void *  buf;
+    uint8_t sbuf[16];
+
+    if (len < sizeof(sbuf)) {
+        return src_read(src, sbuf, len);
+    } else {
+        buf = calloc(1, len);
+
+        if (buf == NULL) {
+            return -1;
+        } else {
+            res = src_read(src, buf, len);
+            free(buf);
+            return res;
+        }
+    }
+}
+
+void
+src_close(pgp_source_t *src)
+{
+    if (src->close) {
+        src->close(src);
+    }
+}
+
+bool
+init_source_cache(pgp_source_t *src, size_t paramsize)
+{
+    if ((src->cache = calloc(1, sizeof(pgp_source_cache_t))) == NULL) {
+        return false;
+    }
+
+    if (paramsize > 0) {
+        if ((src->param = calloc(1, paramsize)) == NULL) {
+            free(src->cache);
+            src->cache = NULL;
+            return false;
+        }
+    } else {
+        src->param = NULL;
+    }
+
+    return true;
+}
+
+typedef struct pgp_source_file_param_t {
+    int fd;
+} pgp_source_file_param_t;
+
+ssize_t
+file_src_read(pgp_source_t *src, void *buf, size_t len)
+{
+    pgp_source_file_param_t *param = src->param;
+
+    if (param == NULL) {
+        return -1;
+    } else {
+        return read(param->fd, buf, len);
+    }
+}
+
+void
+file_src_close(pgp_source_t *src)
+{
+    pgp_source_file_param_t *param = src->param;
+    if (param) {
+        if (src->type == PGP_STREAM_FILE) {
+            close(param->fd);
+        }
+        free(src->param);
+        src->param = NULL;
+    }
+    if (src->cache) {
+        free(src->cache);
+        src->cache = NULL;
+    }
+}
+
+rnp_result_t
+init_file_src(pgp_source_t *src, const char *path)
+{
+    int                      fd;
+    struct stat              st;
+    pgp_source_file_param_t *param;
+
+    if (stat(path, &st) != 0) {
+        (void) fprintf(stderr, "can't stat \"%s\"\n", path);
+        return PGP_E_R_READ_FAILED;
+    }
+
+#ifdef O_BINARY
+    fd = open(path, O_RDONLY | O_BINARY);
+#else
+    fd = open(path, O_RDONLY);
+#endif
+    if (fd < 0) {
+        (void) fprintf(stderr, "can't open \"%s\"\n", path);
+        return RNP_ERROR_READ;
+    }
+
+    if (!init_source_cache(src, sizeof(pgp_source_file_param_t))) {
+        close(fd);
+        return RNP_ERROR_OUT_OF_MEMORY;
+    }
+
+    param = src->param;
+    param->fd = fd;
+    src->read = file_src_read;
+    src->close = file_src_close;
```

Comment:
Just an idea: maybe use a `mmap` here makes the code easy?

Reply:
mmap has some disadvantages, especially for large files: https://stackoverflow.com/questions/258091/when-should-i-use-mmap-for-file-access
Here we just don't need it, since streamed reading is enough.

Reply:
@ni4 at Linux it works a littel different because Linux won't put a mmaped file to swap :) because it is on disk already.

anyway, rnp use mmap at another point and here my idea was make code similar.